src/hamcrest/core/base_matcher.py

```diff
@@ -17,17 +22,17 @@ class BaseMatcher(Matcher):
 
     """
 
-    def __str__(self):
+    def __str__(self) -> str:
         return tostring(self)
 
-    def _matches(self, item):
+    def _matches(self, item: T) -> bool:
```

Comment:
I was thinking it might be helpful for IDE users. They'll see a warning on assertions which cannot possibly pass (assuming the matcher's generic type cas been declared correctly). The concrete matcher can always be declared for `Any` so I don't think it'll be too much of a hardship.

What do you think?

Reply:
I think that the IDE case is nice. I guess it depends on what they're for? I view type annotations as one step short of compiler directives, so I consider what would happen in a static language. Since it's allowable -- and even reasonable -- to have a matcher test against a completely invalid object (it should just fail) I don't see this as the right approach. I'm open to other views on it, though; is anyone else reading this? :)

Reply:
For me, type hints are just that - hints. They never stop you doing anything, they merely want you that you're doing something dodgy, and allow an IDE to offer better assistance.

What's the downside of being more precise here? It'll never stop you writing or running tests - it'll only emit warnings.

Reply:
I suppose we can go with this approach; it's much easier to *loosen* than it is to tighten.